components/object_store/src/disk_cache.rs

```diff
@@ -167,36 +167,63 @@ impl Manifest {
     }
 }
 
-/// The encoder of the page file in the disk cache.
+/// The writer of the page file in the disk cache.
 ///
 /// Following the payload, a footer [`PageFileEncoder::MAGIC_FOOTER`] is
 /// appended.
-struct PageFileEncoder {
-    payload: Bytes,
+struct PageFileWriter {
+    output: String,
 }
 
-impl PageFileEncoder {
+impl PageFileWriter {
     const MAGIC_FOOTER: [u8; 8] = [0, 0, 0, 0, b'c', b'e', b'r', b'e'];
 
-    async fn encode_and_persist<W>(&self, writer: &mut W, name: &str) -> Result<()>
-    where
-        W: AsyncWrite + std::marker::Unpin,
-    {
+    fn new(output: String) -> Self {
+        Self { output }
+    }
+
+    fn tmp_file(input: &str) -> String {
+        format!("{}.tmp", input)
+    }
+
+    async fn write_inner(&self, tmp_file: &str, bytes: Bytes) -> Result<()> {
+        let mut writer = File::create(&tmp_file)
+            .await
+            .context(Io { file: tmp_file })?;
         writer
-            .write_all(&self.payload[..])
+            .write_all(&bytes)
             .await
-            .context(Io { file: name })?;
+            .context(Io { file: tmp_file })?;
 
         writer
             .write_all(&Self::MAGIC_FOOTER)
             .await
-            .context(Io { file: name })?;
+            .context(Io { file: tmp_file })?;
+
+        writer.flush().await.context(Io { file: tmp_file })?;
 
-        writer.flush().await.context(Io { file: name })?;
+        tokio::fs::rename(tmp_file, &self.output)
+            .await
+            .context(Io { file: &self.output })?;
 
         Ok(())
     }
 
+    // When write bytes to file, the cache lock is released, so when one thread is
+    // reading, another thread may update it, so we write to tmp file first,
+    // then rename to expected filename to avoid other threads see partial
+    // content.
+    async fn write_and_flush(self, bytes: Bytes) -> Result<()> {
+        let tmp_file = Self::tmp_file(&self.output);
+        let write_result = self.write_inner(&tmp_file, bytes).await;
+        if write_result.is_err() {
+            // we don't care this result.
+            _ = tokio::fs::remove_file(&tmp_file).await;
+        }
```

Comment:
It's not a better way to remove tempfile. Maybe we should follow RAII style to trigger tempfile removal even if the whole scope (maybe it's tokio task) is dropped.

Reply:
We can't execute async fn in `drop` directly, and I don't want to spawn a runtime inside it.